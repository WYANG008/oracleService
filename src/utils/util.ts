import moment, { DurationInputArg2 } from 'moment';
import WebSocket from 'ws';
import * as CST from '../common/constants';
import { IOption } from '../common/types';

class Util {
	public logLevel: string = CST.LOG_INFO;

	public logInfo(text: any): void {
		this.log(text, CST.LOG_INFO);
	}

	public logDebug(text: any): void {
		this.log(text, CST.LOG_DEBUG);
	}

	public logError(text: any): void {
		this.log(text, CST.LOG_ERROR);
	}

	private log(text: any, level: string): void {
		if (CST.LOG_RANKING[this.logLevel] >= CST.LOG_RANKING[level])
			console.log(`${moment().format('HH:mm:ss.SSS')} [${level}]: ` + text);
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}

	public defaultOption: IOption = {
		source: 'infura',
		provider: ''
	};

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public parseOptions(argv: string[]): IOption {
		const option: IOption = this.defaultOption;
		for (let i = 3; i < argv.length; i++) {
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'source':
					option.source = args[1] || option.source;
					break;
				case 'provider':
					option.provider = args[1] || option.provider;
					break;
				default:
					break;
			}
		}

		return option;
	}

	public round(num: string | number) {
		return +(Math.floor((num + 'e+8') as any) + 'e-8');
	}

	public safeWsSend(ws: WebSocket, message: string) {
		try {
			ws.send(message);
			return true;
		} catch (error) {
			this.logError(error);
			return false;
		}
	}

	public sleep(ms: number) {
		return new Promise(resolve => {
			setTimeout(resolve, ms);
		});
	}

	public clone(obj: object) {
		return JSON.parse(JSON.stringify(obj));
	}

	public getDates(length: number, step: number, stepSize: DurationInputArg2, format: string) {
		const dates: string[] = [];
		const date = moment.utc(this.getUTCNowTimestamp());
		for (let i = 0; i < length; i++) {
			dates.push(date.format(format));
			date.subtract(step, stepSize);
		}
		dates.sort((a, b) => a.localeCompare(b));

		return dates;
	}
}

const util = new Util();
export default util;
