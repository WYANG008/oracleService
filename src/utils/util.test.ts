import util from './util';

test('isNumber() return true for numbers', () => {
	expect(util.isNumber(5)).toBe(true);
	expect(util.isNumber(5.0)).toBe(true);
});

test('isNumber() return true for empty string and null', () => {
	expect(util.isNumber('')).toBe(true);
	expect(util.isNumber(null)).toBe(true);
});

test('isNumber() return true for number strings', () => {
	expect(util.isNumber('5')).toBe(true);
	expect(util.isNumber('5.0')).toBe(true);
});

test('isNumber() return false for other strings', () => {
	expect(util.isNumber('5.0s')).toBe(false);
	expect(util.isNumber('test')).toBe(false);
	expect(util.isNumber('NaN')).toBe(false);
});

test('isNumber() return false for undefined, infinity, NaN', () => {
	expect(util.isNumber(undefined)).toBe(false);
	expect(util.isNumber(Infinity)).toBe(false);
	expect(util.isNumber(NaN)).toBe(false);
});

test('{}, null, undefined is empty', () => {
	expect(util.isEmptyObject({})).toBe(true);
	expect(util.isEmptyObject(null)).toBe(true);
	expect(util.isEmptyObject(undefined)).toBe(true);
});

test('{test: true} is not empty', () => {
	expect(util.isEmptyObject({ test: true })).toBe(false);
});

test('round', () => {
	expect(util.round('12345')).toMatchSnapshot();
	expect(util.round('12345.000')).toMatchSnapshot();
	expect(util.round('12345.1234567')).toMatchSnapshot();
	expect(util.round('12345.123456789')).toMatchSnapshot();
	expect(util.round('0.123456789123456789')).toMatchSnapshot();
	expect(util.round('12345.123456789123456789')).toMatchSnapshot();
});

test('parseOptions', () => {
	const command = [
		'npm',
		'run',
		'tool',
		'live',
		'debug',
		'token=token',
		'amount=123',
		'maker=456',
		'spender=789',
		'dummy=dummy',
		'server'
	];
	expect(util.parseOptions(command)).toMatchSnapshot();
});

test('safeWsSend', () => {
	const ws = {
		send: jest.fn()
	};
	expect(util.safeWsSend(ws as any, 'message')).toBeTruthy();
	ws.send = jest.fn(() => {
		throw new Error('error');
	});
	expect(util.safeWsSend(ws as any, 'message')).toBeFalsy();
});

test('getDates', () => {
	util.getUTCNowTimestamp = jest.fn(() => 1234567890);
	expect(util.getDates(4, 1, 'days', 'YYYY-MM-DD')).toMatchSnapshot();
});
