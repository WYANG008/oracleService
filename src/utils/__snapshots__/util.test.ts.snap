// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDates 1`] = `
Array [
  "1970-01-12",
  "1970-01-13",
  "1970-01-14",
  "1970-01-15",
]
`;

exports[`parseOptions 1`] = `
Object {
  "source": "infura",
}
`;

exports[`round 1`] = `12345`;

exports[`round 2`] = `12345`;

exports[`round 3`] = `12345.1234567`;

exports[`round 4`] = `12345.12345678`;

exports[`round 5`] = `0.12345678`;

exports[`round 6`] = `12345.12345678`;
